oximeter/db/src/lib.rs

```diff
@@ -70,6 +77,230 @@ pub enum Error {
+    pub timeseries_name: TimeseriesName,
+    pub field_schema: Vec<FieldSchema>,
+    pub datum_type: DatumType,
+    pub created: DateTime<Utc>,
+}
+
+impl TimeseriesSchema {
+    /// Return the schema for the given field.
+    pub fn field_schema<S>(&self, name: S) -> Option<&FieldSchema>
+    where
+        S: AsRef<str>,
+    {
+        self.field_schema.iter().find(|field| field.name == name.as_ref())
+    }
+
+    /// Return the target and metric component names for this timeseries
+    pub fn component_names(&self) -> (&str, &str) {
+        self.timeseries_name
+            .split_once(':')
+            .expect("Incorrectly formatted timseries name")
+    }
+}
+
+impl PartialEq for TimeseriesSchema {
+    fn eq(&self, other: &TimeseriesSchema) -> bool {
+        self.timeseries_name == other.timeseries_name
+            && self.datum_type == other.datum_type
+            && self.field_schema == other.field_schema
+    }
+}
+
+impl From<model::DbTimeseriesSchema> for TimeseriesSchema {
+    fn from(schema: model::DbTimeseriesSchema) -> TimeseriesSchema {
+        TimeseriesSchema {
+            timeseries_name: TimeseriesName::try_from(
+                schema.timeseries_name.as_str(),
+            )
+            .expect("Invalid timeseries name in database"),
+            field_schema: schema.field_schema.into(),
+            datum_type: schema.datum_type,
+            created: schema.created,
+        }
+    }
+}
+
+/// The target identifies the resource or component about which metric data is produced.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, JsonSchema)]
+pub struct Target {
+    pub name: String,
+    pub fields: Vec<Field>,
+}
+
+/// The metric identifies the measured aspect or feature of a target.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, JsonSchema)]
+pub struct Metric {
+    pub name: String,
+    pub fields: Vec<Field>,
+    pub datum_type: DatumType,
+}
+
+/// A list of timestamped measurements from a single timeseries.
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, JsonSchema)]
+pub struct Timeseries {
+    pub timeseries_name: String,
+    pub target: Target,
+    pub metric: Metric,
+    pub measurements: Vec<Measurement>,
+}
+
+/// The source from which a field is derived, the target or metric.
+#[derive(
+    Clone,
+    Copy,
+    Debug,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Deserialize,
+    Serialize,
+    JsonSchema,
+)]
+pub enum FieldSource {
+    Target,
+    Metric,
+}
+
+/// The name and type information for a field of a timeseries schema.
+#[derive(
+    Clone,
+    Debug,
+    PartialEq,
+    Eq,
+    PartialOrd,
+    Ord,
+    Deserialize,
+    Serialize,
+    JsonSchema,
+)]
+pub struct FieldSchema {
+    pub name: String,
+    pub ty: FieldType,
+    pub source: FieldSource,
+}
+
+/// Type used to paginate request to list timeseries schema.
+pub type TimeseriesSchemaPaginationParams =
+    PaginationParams<EmptyScanParams, TimeseriesName>;
```

Comment:
Was it intentional that we use `EmptyScanParams` here even though we define scan params that show up in the page token?

Reply:
I don't quite follow. My understanding from the Dropshot documentation is this. The `ScanParams` type is the input we use to start a new scan. The `PageSelector` type is used to resume paginating, from the place we left off. If that's all correct, my intention here was that we have no parameters when starting a scan. We always list from the top. But we can resume a scan from the last collected timeseries schema.

This seems to be exactly how Dropshot's [`example/pagination-basic.rs`](https://github.com/oxidecomputer/dropshot/blob/main/dropshot/examples/pagination-basic.rs) works. Am I missing something about the use of `EmptyScanParams` here?

Reply:
Sorry!  I confused myself thinking this type definition was related to the subsequent scan params and page selector.  I see that this is Timeseries _Schema_ PaginationParams.

Reply:
We talked about this offline, but I think that this was confusion about the pagination parameters for timeseries _schema_ and _data_. For schema, we always start from the first entry, ordered by name, so there are no scan parameters.